app/Actions/AlbumAuthorisationProvider.php

```diff
@@ -0,0 +1,553 @@
+	 * of the origin), the runtime is O(n), but for a high tree (the nodes are
+	 * basically a sequence), the runtime is O(n²).
+	 *
+	 * @param Builder    $query  the album query which shall be restricted
+	 * @param Album|null $origin the optional top album which is used as a search base
+	 *
+	 * @return Builder the restricted album query
+	 */
+	public function applyBrowsabilityFilter(Builder $query, ?Album $origin = null): Builder
+	{
+		$table = $query->getQuery()->from;
+		if (!($query->getModel() instanceof Album) || $table !== 'albums') {
+			throw new \InvalidArgumentException('the given query does not query for albums');
+		}
+
+		// Ensures that only those albums of the original query are
+		// returned for which a path from the origin to the album exist ...
+		if ($origin) {
+			$query
+				// (We include the origin here, because we want the
+				// origin to be browsable from itself)
+				->where('albums._lft', '>=', $origin->_lft)
+				->where('albums._rgt', '<=', $origin->_rgt);
+		}
+
+		// ... such that there are no blocked albums on the path to the album.
+		if (AccessControl::is_admin()) {
+			return $query;
+		} else {
+			return $query->whereNotExists(function (BaseBuilder $q) use ($origin) {
+				$this->appendBlockedAlbumsCondition(
+					$q,
+					$origin ? $origin->_lft : null,
+					$origin ? $origin->_rgt : null,
+				);
+			});
+		}
+	}
+
+	/**
+	 * Adds the conditions of an accessible album to the query.
+	 *
+	 * **Attention:** This method is only meant for internal use by
+	 * this class or {@link PhotoAuthorisationProvider}.
+	 * Use {@link AlbumAuthorisationProvider::applyBrowsabilityFilter()}
+	 * if called from other places instead.
+	 *
+	 * This method adds the WHERE conditions without any further pre-cautions.
+	 * The method silently assumes that the passed query builder is used
+	 * within an outer query whose SELECT clause contains the table
+	 *
+	 *  - **`albums`**.
+	 *
+	 * Moreover, the raw clauses are added.
+	 * They are not wrapped into a nesting braces `()`.
+	 *
+	 * @param BaseBuilder     $builder     the album query which shall be
+	 *                                     restricted
+	 * @param int|string|null $originLeft  optionally constraints the search
+	 *                                     base; an integer value is
+	 *                                     interpreted a raw left bound of the
+	 *                                     search base; a string value is
+	 *                                     interpreted as a reference to a
+	 *                                     column which shall be used as a
+	 *                                     left bound
+	 * @param int|string|null $originRight like `$originLeft` but for the
+	 *                                     right bound
+	 *
+	 * @return BaseBuilder
+	 *
+	 * @throws \InvalidArgumentException
+	 */
+	public function appendBlockedAlbumsCondition(BaseBuilder $builder, $originLeft, $originRight): BaseBuilder
```

Comment:
Same as above.
https://stitcher.io/blog/new-in-php-8#union-types-rfc

Reply:
I know, I am looking forward to it.